File: webapp/server/repositories/expenseRepository.ts
import { FieldValue, type Timestamp } from "firebase-admin/firestore";
import type { Expense } from "@/shared/types/group";
import { getAdminFirestore } from "../lib/firebaseAdmin";

export type ExpenseRecord = {
  groupId: string;
  description: string;
  amount: number;
  paidBy: string;
  splitWith: string[];
};

export async function createExpense(
  record: ExpenseRecord
): Promise<{ id: string }> {
  const db = getAdminFirestore();
  const docRef = await db
    .collection("groups")
    .doc(record.groupId)
    .collection("expenses")
    .add({
      description: record.description,
      amount: record.amount,
      paidBy: record.paidBy,
      splitWith: record.splitWith,
      createdAt: FieldValue.serverTimestamp(),
      updatedAt: FieldValue.serverTimestamp(),
    });
  return { id: docRef.id };
}

export async function getExpenses(groupId: string): Promise<Expense[]> {
  const db = getAdminFirestore();
  const snapshot = await db
    .collection("groups")
    .doc(groupId)
    .collection("expenses")
    .orderBy("createdAt", "desc")
    .get();

  return snapshot.docs.map((doc) => {
    const data = doc.data();
    return {
      id: doc.id,
      groupId,
      description: data.description,
      amount: data.amount,
      paidBy: data.paidBy,
      splitWith: data.splitWith,
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    };
  });
}

export async function deleteExpense(
  groupId: string,
  expenseId: string
): Promise<void> {
  const db = getAdminFirestore();
  await db
    .collection("groups")
    .doc(groupId)
    .collection("expenses")
    .doc(expenseId)
    .delete();
}

File: mobileapp/lib/repositories/expense_repository.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/expense.dart';

class ExpenseRepository {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  /// Get all expenses for a group
  Future<List<Expense>> getExpenses(String groupId) async {
    final snapshot = await _firestore
        .collection('groups')
        .doc(groupId)
        .collection('expenses')
        .orderBy('createdAt', descending: true)
        .get();

    return snapshot.docs
        .map((doc) => Expense.fromFirestore(doc, groupId))
        .toList();
  }

  /// Create a new expense
  Future<String> createExpense(
    String groupId, {
    required String description,
    required double amount,
    required String paidBy,
    required List<String> splitWith,
  }) async {
    final docRef = await _firestore
        .collection('groups')
        .doc(groupId)
        .collection('expenses')
        .add({
      'description': description,
      'amount': amount,
      'paidBy': paidBy,
      'splitWith': splitWith,
      'createdAt': FieldValue.serverTimestamp(),
      'updatedAt': FieldValue.serverTimestamp(),
    });

    // Update group's updatedAt
    await _firestore.collection('groups').doc(groupId).update({
      'updatedAt': FieldValue.serverTimestamp(),
    });

    return docRef.id;
  }

  /// Delete an expense
  Future<void> deleteExpense(String groupId, String expenseId) async {
    await _firestore
        .collection('groups')
        .doc(groupId)
        .collection('expenses')
        .doc(expenseId)
        .delete();

    // Update group's updatedAt
    await _firestore.collection('groups').doc(groupId).update({
      'updatedAt': FieldValue.serverTimestamp(),
    });
  }
}
